### ISSUE TITLE
can't get tracking working in a react-native native component 

### ISSUE THREAD

**Issue by matthieugayon** [at-internet/atinternet-android-sdk#7]
Here is my implementation using the react-native android bridge abilities, unfortunately it is not sending any hits, it is compiling fine but nothing gets sent,  any help would be much appreciated :)

TrackerModule.java 

```java
package com.gallicadabra.tracker;

import android.app.Activity;
import android.util.Log;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;

import java.util.HashMap;
import java.util.Map;
import java.util.ArrayList;

import com.atinternet.tracker.ATInternet;
import com.atinternet.tracker.Tracker;
import com.atinternet.tracker.CustomVar;

import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;

public class TrackerModule extends ReactContextBaseJavaModule {

    private Tracker tracker;

    public TrackerModule(ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @Override
    public String getName() {
        return "Xiti";
    }

    @ReactMethod
    public void initTracker(String trackerName) {
        //final Activity activity = getCurrentActivity();

        tracker = new Tracker(this.getReactApplicationContext(), new HashMap<String, Object>() {{
            put("log","logp");
            put("logSSL", "logs");
            put("domain", "xiti.com");
            put("pixelPath", "/hit.xiti");
            put("site", 18798);
            put("secure", true);
            put("identifier", "uuid");
            put("enableCrashDetection", true);
            put("plugins", "tvtracking");
            put("storage", "required");
            put("hashUserId", false);
            put("persistIdentifiedVisitor", true);
            put("tvtURL", "");
            put("tvtVisitDuration", 10);
            put("campaignLastPersistence", false);
            put("campaignLifetime", 30);
            put("sessionBackgroundDuration", 60);
        }});
    }

    @ReactMethod
    public void sendView(String viewName, ReadableArray customVars, ReadableMap chapters) {

        tracker.Context().setLevel2(73);

        for (int i = 0; i < customVars.size(); i++) {
            ReadableMap customVar = customVars.getMap(i);
            int id = customVar.hasKey("id") ? customVar.getInt("id") : 0;
            String value = customVar.hasKey("value") ? customVar.getString("value") : "";
            tracker.CustomVars().add(id, value, CustomVar.CustomVarType.App);
        }

        String chapter1 = chapters.hasKey("chapter1") ? chapters.getString("chapter1") : null;
        String chapter2 = chapters.hasKey("chapter2") ? chapters.getString("chapter2") : null;
        String chapter3 = chapters.hasKey("chapter3") ? chapters.getString("chapter3") : null;

        Log.v("chapter1", chapter1 + "");
        Log.v("chapter2", chapter2 + "");
        Log.v("chapter3", chapter3 + "");


        tracker.Screens()
                .add(viewName, chapter1, chapter2, chapter3)
                .setLevel2(73);
        tracker.dispatch();
    }

    @ReactMethod
    public void sendTouch(String touchName, ReadableMap chapters) {

        tracker.Context().setLevel2(73);

        String chapter1 = chapters.hasKey("chapter1") ? chapters.getString("chapter1") : null;
        String chapter2 = chapters.hasKey("chapter2") ? chapters.getString("chapter2") : null;
        String chapter3 = chapters.hasKey("chapter3") ? chapters.getString("chapter3") : null;


        tracker.Gestures()
                .add(touchName, chapter1, chapter2, chapter3)
                .setLevel2(73)
                .sendTouch();
    }

    @ReactMethod
    public void sendNavigation(String eventName, ReadableMap chapters) {

        tracker.Context().setLevel2(73);

        String chapter1 = chapters.hasKey("chapter1") ? chapters.getString("chapter1") : null;
        String chapter2 = chapters.hasKey("chapter2") ? chapters.getString("chapter2") : null;
        String chapter3 = chapters.hasKey("chapter3") ? chapters.getString("chapter3") : null;


        tracker.Gestures()
                .add(eventName, chapter1, chapter2, chapter3)
                .setLevel2(73)
                .sendNavigation();
    }

    @ReactMethod
    public void sendOfflineHits() {

        int nbHit = tracker.Offline().count();

        if (nbHit > 0) {
            tracker.Offline().dispatch();
        }
    }
}
```

TrackerPackage.java 

```java
package com.gallicadabra.tracker;

import com.facebook.react.ReactPackage;
import com.facebook.react.bridge.JavaScriptModule;
import com.facebook.react.bridge.NativeModule;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.uimanager.ViewManager;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class TrackerPackage implements ReactPackage {

    @Override
    public List<Class<? extends JavaScriptModule>> createJSModules() {
        return Collections.emptyList();
    }

    @Override
    public List<ViewManager> createViewManagers(ReactApplicationContext reactContext) {
        return Collections.emptyList();
    }

    @Override
    public List<NativeModule> createNativeModules(
            ReactApplicationContext reactContext) {
        List<NativeModule> modules = new ArrayList<>();

        modules.add(new TrackerModule(reactContext));

        return modules;
    }
}
```

MainActivity.java

```java
package com.gallicadabra;

import com.facebook.react.ReactActivity;

// react-native-orientation
import android.content.Intent;
import android.content.res.Configuration;

public class MainActivity extends ReactActivity {

    /**
     * Returns the name of the main component registered from JavaScript.
     * This is used to schedule rendering of the component.
     */
    @Override
    protected String getMainComponentName() {
        return "Gallicadabra";
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        Intent intent = new Intent("onConfigurationChanged");
        intent.putExtra("newConfig", newConfig);
        this.sendBroadcast(intent);
    }
}
```

MainApplication.java 

```java
package com.gallicadabra;

import android.app.Application;
import android.util.Log;

import com.facebook.react.ReactApplication;
import com.gallicadabra.tracker.TrackerPackage;
import com.github.yamill.orientation.OrientationPackage;
import com.reactnative.photoview.PhotoViewPackage;
import com.zmxv.RNSound.RNSoundPackage;
import com.facebook.react.ReactInstanceManager;
import com.facebook.react.ReactNativeHost;
import com.facebook.react.ReactPackage;
import com.facebook.react.shell.MainReactPackage;

import java.util.Arrays;
import java.util.List;

public class MainApplication extends Application implements ReactApplication {

  private final ReactNativeHost mReactNativeHost = new ReactNativeHost(this) {

    @Override
    protected boolean getUseDeveloperSupport() {
      return BuildConfig.DEBUG;
    }

    @Override
    protected List<ReactPackage> getPackages() {
      return Arrays.<ReactPackage>asList(
          new MainReactPackage(),
          new OrientationPackage(),
          new PhotoViewPackage(),
          new RNSoundPackage(),
          new TrackerPackage()
      );
    }
  };

  @Override
  public ReactNativeHost getReactNativeHost() {
      return mReactNativeHost;
  }
}
```

AndroidManifest.xml

```xml
<manifest xmlns:android="http://schemas.android.com/apk/res/android"
    package="com.gallicadabra"
    android:versionCode="1"
    android:versionName="1.0">

    <uses-permission android:name="android.permission.INTERNET" />
    <uses-permission android:name="android.permission.SYSTEM_ALERT_WINDOW"/>
    <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />

    <uses-sdk
        android:minSdkVersion="16"
        android:targetSdkVersion="22" />

    <application
      android:name=".MainApplication"
      android:allowBackup="true"
      android:label="@string/app_name"
      android:icon="@mipmap/ic_launcher"
      android:theme="@style/AppTheme">
      <activity
        android:name=".MainActivity"
        android:label="@string/app_name"
        android:configChanges="keyboard|keyboardHidden|orientation|screenSize">
        <intent-filter>
            <action android:name="android.intent.action.MAIN" />
            <category android:name="android.intent.category.LAUNCHER" />
        </intent-filter>
      </activity>
      <activity android:name="com.facebook.react.devsupport.DevSettingsActivity" />
    </application>

</manifest>
```

**Comment by nsagnett** [comment]
HI !

It's just because in your configuration, you specify "uuid" identifier type.

But configuration is case sensitive. Try with "UUID" instead.

Regards

**Comment by matthieugayon** [comment]
re opened ?

**Comment by nsagnett** [comment]
Yes I'm waiting for your feedback to close issue.

**Comment by matthieugayon** [comment]
I'll check that out and get back to you later today when I find the time, thanks for your help anyway !

**Comment by matthieugayon** [comment]
It seems it was not problem, I setup a trackerListener like in the initTracker Method, and you can see a screenshot of the logs from the android studio monitor bellow, every time the SendView if called the chapters values are being printed but the trackerListener doesn't log any info ... Any thought ? 

<img width="556" alt="screen shot 2017-04-04 at 18 18 01" src="https://cloud.githubusercontent.com/assets/3694172/24667398/9eccbf74-1963-11e7-9f23-031d65d27a76.png">

 
```java
package com.gallicadabra.tracker;

import android.app.Activity;
import android.util.Log;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;

import java.util.HashMap;
import java.util.Map;
import java.util.ArrayList;

import com.atinternet.tracker.ATInternet;
import com.atinternet.tracker.Tracker;
import com.atinternet.tracker.CustomVar;
import com.atinternet.tracker.TrackerListener;

import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;

public class TrackerModule extends ReactContextBaseJavaModule {

    private Tracker tracker;

    public TrackerModule(ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @Override
    public String getName() {
        return "Xiti";
    }

    @ReactMethod
    public void initTracker(String trackerName) {
        //final Activity activity = getCurrentActivity();

        tracker = new Tracker(this.getReactApplicationContext(), new HashMap<String, Object>() {{
            put("log","logp");
            put("logSSL", "logs");
            put("domain", "xiti.com");
            put("pixelPath", "/hit.xiti");
            put("site", 18798);
            put("secure", true);
            put("identifier", "UUID");
            put("enableCrashDetection", true);
            put("plugins", "tvtracking");
            put("storage", "required");
            put("hashUserId", false);
            put("persistIdentifiedVisitor", true);
            put("tvtURL", "");
            put("tvtVisitDuration", 10);
            put("campaignLastPersistence", false);
            put("campaignLifetime", 30);
            put("sessionBackgroundDuration", 60);
        }});

        TrackerListener screenListener = new TrackerListener() {
            @Override
            public void trackerNeedsFirstLaunchApproval(String message) {
                String flag = "NeedsFrstLaunchAppr";
                Log.d(flag, message);
            }

            @Override
            public void buildDidEnd(HitStatus status, String message) {
                String flag = "buildDidEnd";
                Log.d(flag, message);
            }

            @Override
            public void sendDidEnd(HitStatus status, String message) {
                String flag = "sendDidEnd";
                Log.d(flag, message);
            }

            @Override
            public void didCallPartner(String response) {

            }

            @Override
            public void warningDidOccur(String message) {
                String flag = "warningDidOccur";
                Log.d(flag, message);
            }

            @Override
            public void saveDidEnd(String message) {
                String flag = "saveDidEnd";
                Log.d(flag, message);
            }

            @Override
            public void errorDidOccur(String message) {
                String flag = "errorDidOccur";
                Log.d(flag, message);
            }
        };

        tracker.setListener(screenListener);
    }

    @ReactMethod
    public void sendView(String viewName, ReadableArray customVars, ReadableMap chapters) {

        tracker.Context().setLevel2(73);

        for (int i = 0; i < customVars.size(); i++) {
            ReadableMap customVar = customVars.getMap(i);
            int id = customVar.hasKey("id") ? customVar.getInt("id") : 0;
            String value = customVar.hasKey("value") ? customVar.getString("value") : "";
            tracker.CustomVars().add(id, value, CustomVar.CustomVarType.App);
        }

        String chapter1 = chapters.hasKey("chapter1") ? chapters.getString("chapter1") : null;
        String chapter2 = chapters.hasKey("chapter2") ? chapters.getString("chapter2") : null;
        String chapter3 = chapters.hasKey("chapter3") ? chapters.getString("chapter3") : null;

        Log.v("chapter1", chapter1 + "");
        Log.v("chapter2", chapter2 + "");
        Log.v("chapter3", chapter3 + "");


        tracker.Screens()
                .add(viewName, chapter1, chapter2, chapter3)
                .setLevel2(73)
                .sendView();
    }

    @ReactMethod
    public void sendTouch(String touchName, ReadableMap chapters) {

        tracker.Context().setLevel2(73);

        String chapter1 = chapters.hasKey("chapter1") ? chapters.getString("chapter1") : null;
        String chapter2 = chapters.hasKey("chapter2") ? chapters.getString("chapter2") : null;
        String chapter3 = chapters.hasKey("chapter3") ? chapters.getString("chapter3") : null;


        tracker.Gestures()
                .add(touchName, chapter1, chapter2, chapter3)
                .setLevel2(73)
                .sendTouch();
    }

    @ReactMethod
    public void sendNavigation(String eventName, ReadableMap chapters) {

        tracker.Context().setLevel2(73);

        String chapter1 = chapters.hasKey("chapter1") ? chapters.getString("chapter1") : null;
        String chapter2 = chapters.hasKey("chapter2") ? chapters.getString("chapter2") : null;
        String chapter3 = chapters.hasKey("chapter3") ? chapters.getString("chapter3") : null;


        tracker.Gestures()
                .add(eventName, chapter1, chapter2, chapter3)
                .setLevel2(73)
                .sendNavigation();
    }

    @ReactMethod
    public void sendOfflineHits() {

        int nbHit = tracker.Offline().count();

        if (nbHit > 0) {
            tracker.Offline().dispatch();
        }
    }
}

```

**Comment by nsagnett** [comment]
The problem occured again ? I copied your code but cannot reproduce your issue, logs are shown..

**Comment by matthieugayon** [comment]
I solved my problem by replacing the line 

put("identifier", "UUID");
with 
put("identifier", "androidId");

as simple as that, I believe this should be explained in the doc ...

**Comment by nsagnett** [comment]
What's your SDK version ?

**Comment by matthieugayon** [comment]
2.0.6

**Comment by nsagnett** [comment]
Ah yes okay it's the problem. If you check the changelog (https://developers.atinternet-solutions.com/android-en/changelog-android-en/), UUID is available since 2.3.2 :).
